From: Hannes Frederic Sowa <hannes@stressinduktion.org>
Date: Mon, 6 Oct 2014 19:58:37 +0200
Subject: ipv6: make fib6 serial number per namespace
Patch-mainline: v3.18-rc1
Git-commit: 812918c464eca0e8c145f975932ca5020e9c05cb
References: bsc#965319

Try to reduce number of possible fn_sernum mutation by constraining them
to their namespace.

Also remove rt_genid which I forgot to remove in 705f1c869d577c ("ipv6:
remove rt6i_genid").

Cc: YOSHIFUJI Hideaki <hideaki@yoshifuji.org>
Cc: Martin Lau <kafai@fb.com>
Signed-off-by: Hannes Frederic Sowa <hannes@stressinduktion.org>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Michal Kubecek <mkubecek@suse.cz>

SLE11-SP3: no rt_genid removal

---
 include/net/netns/ipv6.h |  1 +
 net/ipv6/ip6_fib.c       | 23 ++++++++++++-----------
 2 files changed, 13 insertions(+), 11 deletions(-)

diff --git a/include/net/netns/ipv6.h b/include/net/netns/ipv6.h
index aeea97f6b7e1..de3d31b8f029 100644
--- a/include/net/netns/ipv6.h
+++ b/include/net/netns/ipv6.h
@@ -49,6 +49,7 @@ struct netns_ipv6 {
 	struct fib6_table       *fib6_main_tbl;
 	struct dst_ops		ip6_dst_ops;
 	unsigned int		 ip6_rt_gc_expire;
+	__u32			fib6_sernum;
 	unsigned long		 ip6_rt_last_gc;
 #ifdef CONFIG_IPV6_MULTIPLE_TABLES
 	struct rt6_info         *ip6_prohibit_entry;
diff --git a/net/ipv6/ip6_fib.c b/net/ipv6/ip6_fib.c
index 80a97937a4e8..a6165569b11c 100644
--- a/net/ipv6/ip6_fib.c
+++ b/net/ipv6/ip6_fib.c
@@ -91,8 +91,6 @@ static int fib6_walk_continue(struct fib6_walker_t *w);
  *	result of redirects, path MTU changes, etc.
  */
 
-static __u32 rt_sernum;
-
 static void fib6_gc_timer_cb(unsigned long arg);
 
 static LIST_HEAD(fib6_walkers);
@@ -111,11 +109,11 @@ static inline void fib6_walker_unlink(struct fib6_walker_t *w)
 	list_del(&w->lh);
 	write_unlock_bh(&fib6_walker_lock);
 }
-static __inline__ u32 fib6_new_sernum(void)
+static __inline__ u32 fib6_new_sernum(struct net *net)
 {
-	u32 n = ++rt_sernum;
+	u32 n = ++(net->ipv6.fib6_sernum);
 	if ((__s32)n <= 0)
-		rt_sernum = n = 1;
+		net->ipv6.fib6_sernum = n = 1;
 	return n;
 }
 
@@ -431,14 +429,14 @@ out:
 static struct fib6_node * fib6_add_1(struct fib6_node *root, void *addr,
 				     int addrlen, int plen,
 				     int offset, int allow_create,
-				     int replace_required)
+				     int replace_required, struct net *net)
 {
 	struct fib6_node *fn, *in, *ln;
 	struct fib6_node *pn = NULL;
 	struct rt6key *key;
 	int	bit;
 	__be32	dir = 0;
-	__u32	sernum = fib6_new_sernum();
+	__u32	sernum = fib6_new_sernum(net);
 
 	RT6_TRACE("fib6_add_1\n");
 
@@ -778,6 +776,8 @@ int fib6_add(struct fib6_node *root, struct rt6_info *rt, struct nl_info *info,
 	int err = -ENOMEM;
 	int allow_create = 1;
 	int replace_required = 0;
+	struct net *net = info->nl_net;
+
 	if (NULL != info && NULL != info->nlh) {
 		if (!(info->nlh->nlmsg_flags&NLM_F_CREATE))
 			allow_create = 0;
@@ -789,7 +789,7 @@ int fib6_add(struct fib6_node *root, struct rt6_info *rt, struct nl_info *info,
 
 	fn = fib6_add_1(root, &rt->rt6i_dst.addr, sizeof(struct in6_addr),
 		    rt->rt6i_dst.plen, offsetof(struct rt6_info, rt6i_dst),
-		    allow_create, replace_required);
+		    allow_create, replace_required, net);
 
 	if (IS_ERR(fn)) {
 		err = PTR_ERR(fn);
@@ -826,14 +826,14 @@ int fib6_add(struct fib6_node *root, struct rt6_info *rt, struct nl_info *info,
 			sfn->leaf = info->nl_net->ipv6.ip6_null_entry;
 			atomic_inc(&info->nl_net->ipv6.ip6_null_entry->rt6i_ref);
 			sfn->fn_flags = RTN_ROOT;
-			sfn->fn_sernum = fib6_new_sernum();
+			sfn->fn_sernum = fib6_new_sernum(net);
 
 			/* Now add the first leaf node to new subtree */
 
 			sn = fib6_add_1(sfn, &rt->rt6i_src.addr,
 					sizeof(struct in6_addr), rt->rt6i_src.plen,
 					offsetof(struct rt6_info, rt6i_src),
-					allow_create, replace_required);
+					allow_create, replace_required, net);
 
 			if (sn == NULL) {
 				/* If it is failed, discard just allocated
@@ -851,7 +851,7 @@ int fib6_add(struct fib6_node *root, struct rt6_info *rt, struct nl_info *info,
 			sn = fib6_add_1(fn->subtree, &rt->rt6i_src.addr,
 					sizeof(struct in6_addr), rt->rt6i_src.plen,
 					offsetof(struct rt6_info, rt6i_src),
-					allow_create, replace_required);
+					allow_create, replace_required, net);
 
 			if (IS_ERR(sn)) {
 				err = PTR_ERR(sn);
@@ -1613,6 +1613,7 @@ static int __net_init fib6_net_init(struct net *net)
 {
 	size_t size = sizeof(struct hlist_head) * FIB6_TABLE_HASHSZ;
 
+	net->ipv6.fib6_sernum = 1;
 	setup_timer(&net->ipv6.ip6_fib_timer, fib6_gc_timer_cb, (unsigned long)net);
 
 	net->ipv6.rt6_stats = kzalloc(sizeof(*net->ipv6.rt6_stats), GFP_KERNEL);
-- 
2.8.1

